### PR TITLE
Architecture: Scrypto Env

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: radix-engine-lib/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+[package]
+name = "radix-engine-lib"
```
Comment (iamyulong) [inline radix-engine-lib/Cargo.toml:2]: Rename to `radix-engine-interface`?

File: radix-engine-lib/src/core/mod.rs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+pub mod expression;
```
Comment (iamyulong) [inline radix-engine-lib/src/core/mod.rs:1]: Can we move all types in this folder to `engine/types`?

File: radix-engine-lib/src/data/indexed_value.rs [hunk 3]
```diff
@@ -332,11 +331,16 @@ mod tests {
     use super::rust::vec;
     use super::*;
 
+    /// Encodes a data structure into byte array.
+    fn scrypto_encode<T: Encode<ScryptoCustomTypeId> + ?Sized>(v: &T) -> Vec<u8> {
```
Comment (iamyulong) [inline radix-engine-lib/src/data/indexed_value.rs:335]: Would consider adding this to `data/mod.rs`. And the decode.

File: radix-engine-lib/src/engine/api.rs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+pub trait SysInvocation {
+    type Output: Debug + Decode<ScryptoCustomTypeId>;
+}
+
+pub trait SysNativeInvokable<I: SysInvocation, E> {
+    fn sys_invoke(&mut self, invocation: I) -> Result<I::Output, E>;
+}
+
+pub trait Syscalls<E: Debug> {
+    fn sys_invoke_scrypto_function(
```
Comment (iamyulong) [inline radix-engine-lib/src/engine/api.rs:24]: Can we get rid of all these `sys_` prefix?

File: radix-engine-lib/src/engine/actor.rs [hunk 5]
```diff
@@ -1,7 +1,7 @@
 use sbor::rust::string::String;
 
-use crate::component::*;
 use crate::engine::types::ComponentId;
+use crate::model::*;
 use crate::scrypto;
 
 #[derive(Debug, Clone)]
```
Comment (iamyulong) [inline radix-engine-lib/src/engine/actor.rs:7]: Can we move this file to `engine/types`?

File: radix-engine-lib/src/engine/api.rs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+use sbor::rust::fmt::Debug;
+use sbor::rust::string::String;
+use sbor::rust::vec::Vec;
+use sbor::Decode;
+
+use super::types::*;
+
+pub trait ScryptoNativeInvocation: Into<NativeFnInvocation> + SysInvocation {}
+
+pub trait SysInvocation {
```
Comment (iamyulong) [inline radix-engine-lib/src/engine/api.rs:15]: We can remove the `Sys` prefix, so we have:
- `Invocation`
- `ScryptoNativeInvocation`
- `NativeInvokable`
- `NativeInvokables` - the set of invkables

File: radix-engine-lib/src/engine/api.rs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    type Output: Debug + Decode<ScryptoCustomTypeId>;
+}
+
+pub trait SysNativeInvokable<I: SysInvocation, E> {
+    fn sys_invoke(&mut self, invocation: I) -> Result<I::Output, E>;
+}
+
+pub trait Syscalls<E: Debug> {
```
Comment (iamyulong) [inline radix-engine-lib/src/engine/api.rs:23]: Would rename to `EngineApi`

File: radix-engine-lib/src/engine/scrypto_env.rs [hunk 8]
```diff
@@ -0,0 +1,167 @@
+use crate::engine::types::*;
```
Comment (iamyulong) [inline radix-engine-lib/src/engine/scrypto_env.rs:1]: Would rename to `api_wasm.rs`.

File: radix-engine-lib/src/engine/types/third_party.rs [hunk 9]
```diff
@@ -6,8 +6,5 @@ pub use sbor::*;
 pub use sbor::*;
 pub use strum::*;
 
-pub use crate::component::{ComponentAddress, PackageAddress};
-pub use crate::core::SystemAddress;
 pub use crate::crypto::Hash;
-pub use crate::resource::NonFungibleId;
-pub use crate::resource::ResourceAddress;
+pub use crate::model::*;
```
Comment (iamyulong) [inline radix-engine-lib/src/engine/types/third_party.rs:10]: Maybe it's time to get rid of this file, as we have a dedicated engine interface crate.

File: radix-engine-lib/src/macros.rs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+                base / radix_engine_lib::math::PreciseDecimal::try_from(
+                    radix_engine_lib::math::I512::from(10u8)
+                        .pow(u32::try_from(-$shift).expect("Shift overflow")),
+                )
+                .expect("Shift overflow")
+            }
+        }
+    };
+}
+
```
Comment (iamyulong) [inline radix-engine-lib/src/macros.rs:71]: Can we live without the above integer and decimal macros? Don't want to maintain two copy of code and they're mainly for developer's convenience.

File: radix-engine-lib/src/math/integer/bits.rs [hunk 11]
```diff
@@ -13,17 +13,6 @@ macro_rules! checked_int_impl_large {
...
-                    ///
-                    /// Basic usage:
-                    ///
-                    /// ```
-                    #[doc = "use scrypto::prelude::*;"]
-                    ///
-                    #[doc = "let n = " $t "::by(0b01001100u8);"]
-                    ///
-                    /// assert_eq!(n.count_ones(), 3);
-                    /// ```
```
Comment (iamyulong) [inline radix-engine-lib/src/math/integer/bits.rs:26]: Possibly to make them compile? The documentation will be re-exported to scrypto and might be useful for developers. Also, they're also used for testing if I'm not mistaken.

File: radix-engine-lib/src/model/resource/non_fungible_id.rs [hunk 12]
```diff
@@ -7,22 +7,19 @@ use sbor::rust::vec::Vec;
 use sbor::*;
 
 use crate::abi::*;
-use crate::buffer::scrypto_encode;
-use crate::data::IndexedScryptoValue;
 use crate::data::*;
 use crate::scrypto_type;
 
+/// Encodes a data structure into byte array.
+fn scrypto_encode<T: Encode<ScryptoCustomTypeId> + ?Sized>(v: &T) -> Vec<u8> {
```
Comment (iamyulong) [inline radix-engine-lib/src/model/resource/non_fungible_id.rs:14]: Again.

File: radix-engine-lib/src/model/component/data.rs [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+use crate::abi::*;
+use crate::address::*;
+use crate::crypto::{hash, PublicKey};
+use crate::data::ScryptoCustomTypeId;
+use crate::scrypto_type;
+
+/// An instance of a blueprint, which lives in the ledger state.
+#[derive(Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum ComponentAddress {
```
Comment (iamyulong) [inline radix-engine-lib/src/model/component/data.rs:15]: Would be nice if we can have all global addresses under `model/global`.

File: scrypto/src/buffer/codec.rs [hunk 14]
```diff
@@ -27,6 +27,12 @@ pub fn scrypto_decode_from_buffer<T: Decode<ScryptoCustomTypeId> + ?Sized>(
     scrypto_consume(ptr, |slice| scrypto_decode(slice))
 }
 
+/// Decode a data structure from a Scrypto buffer.
+pub fn scrypto_raw_from_buffer(ptr: *mut u8) -> Vec<u8> {
```
Comment (iamyulong) [inline scrypto/src/buffer/codec.rs:31]: `scrypto_buffer_to_vec`?

File: scrypto/src/component/component.rs [hunk 15]
```diff
@@ -95,22 +95,40 @@ impl Component {
...
-                access_rules,
-                receiver: self.0,
-            }),
-        );
-        let _: () = call_engine(input);
+        self.sys_add_access_check(access_rules, &mut ScryptoEnv)
+            .unwrap()
+    }
+
+    pub fn sys_add_access_check<Y, E: Debug + Decode<ScryptoCustomTypeId>>(
```
Comment (iamyulong) [inline scrypto/src/component/component.rs:102]: The `sys_` variant is no longer needed right?

File: scrypto/src/component/component.rs [hunk 16]
```diff
@@ -95,22 +95,40 @@ impl Component {
...
     }
 
     pub fn globalize(self) -> ComponentAddress {
-        let input = RadixEngineInput::CreateNode(ScryptoRENode::GlobalComponent(self.0));
-        let node_id: RENodeId = call_engine(input);
-        node_id.into()
+        self.sys_globalize(&mut ScryptoEnv).unwrap()
+    }
+
+    pub fn sys_globalize<Y, E: Debug + Decode<ScryptoCustomTypeId>>(
```
Comment (iamyulong) [inline scrypto/src/component/component.rs:122]: Same

File: scrypto/src/core/runtime.rs [hunk 17]
```diff
@@ -1,44 +1,45 @@
...
 /// The transaction runtime.
 #[derive(Debug)]
 pub struct Runtime {}
 
 impl Runtime {
+    pub fn current_epoch() -> u64 {
+        Self::sys_current_epoch(&mut ScryptoEnv).unwrap()
+    }
+
+    pub fn sys_current_epoch<Y, E>(env: &mut Y) -> Result<u64, E>
```
Comment (iamyulong) [inline scrypto/src/core/runtime.rs:28]: Same

File: scrypto/src/resource/auth_zone.rs [hunk 18]
```diff
@@ -60,108 +17,166 @@ pub struct AuthZoneDrainInvocation {
...
-        );
-        call_engine(input)
+
+        env.sys_invoke(AuthZonePushInvocation {
+            receiver: node_id.into(),
+            proof,
+        })
+    }
+
+    pub fn pop() -> Proof {
```
Comment (iamyulong) [inline scrypto/src/resource/auth_zone.rs:160]: Can remove all the `sys_` variants now?

File: scrypto/src/resource/proof.rs [hunk 19]
```diff
@@ -71,20 +46,82 @@ impl From<NonFungibleAddress> for ProofValidationMode {
...
-        NativeMethod::Proof => {
-            fn clone(&self) -> Self {
-                ProofMethod::Clone,
-                ProofCloneInvocation {
-                    receiver: self.0
-                }
-            }
-        }
+pub trait SysProof {
+    fn sys_clone<Y, E: Debug + TypeId<ScryptoCustomTypeId> + Decode<ScryptoCustomTypeId>>(
```
Comment (iamyulong) [inline scrypto/src/resource/proof.rs:50]: Looks like all these traits are intentional after the refactoring. Will skip for now and resort to a discussion.

**Comment by iamyulong** [comment]
Pushed the first batch of comments. Will re-review the `scrypto` crate in greater details after comments are addressed.

File: radix-engine-lib/src/core/mod.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine-lib/src/core/mod.rs:1]: `network` and `expression` aren't engine related abstractions so would rather just keep them here until we find a more appropriate place to put these.

**Review by iamyulong — APPROVED** [review]

**Pull request merged**